### PR TITLE
Adding Non-Uniform Fast Fourier Transform

### PR DESCRIPTION
**Pull request by bensarthou**
The PR implements the interface between Pysap and PyNUFFT, a Python implementation of the NUFFT, supporting multi-CPU and GPU acceleration. 

As the package is only developed for Python3.x, the dependency is listed as optional and the NUFFT transformation isn't tested (but a commented code is provided in the unittest to do so). 
Furthermore, as installing the correct platform for mCPU/GPU computation is non trivial, the unitest only test the CPU implementation of the NUFFT. 
For mCPU/GPU acceleration, please refer to the pynufft documentation ([http://jyhmiinlin.github.io/pynufft/](url) for more information)

Aside of that, the PR also implement the 3D version of NFFT (and some utils function nedded for the tests)

### CODE REVIEW

File: pysap/plugins/mri/reconstruct_3D/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+##########################################################################
+# XXX - Copyright (C) XXX, 2017
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/__init__.py]: To update.

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
+# -*- coding: utf-8 -*-
+##########################################################################
+# XXX - Copyright (C) XXX, 3017
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: To update.

File: pysap/plugins/mri/reconstruct_3D/utils.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+##########################################################################
+# XXX - Copyright (C) XXX, 2017
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/utils.py]: To update.

File: pysap/scripts/pysap_gridsearch [hunk 4]
```diff
@@ -18,7 +18,7 @@ import json
 import pickle
```
Comment (AGrigis) [inline pysap/scripts/pysap_gridsearch]: Already integrated in master.

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+"""
+Fourier operators for cartesian and non-cartesian space.
+"""
+
+
+# Package import
+import warnings
+from .utils import convert_locations_to_mask_3D
+from pysap.plugins.mri.reconstruct.fourier import FourierBase
+from modopt.interface.errors import warn
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: Why using the modopt warn?
Can we use the one from the warnings package?

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+# Third party import
+try:
+    from pynufft import NUFFT_hsa
+    from pynufft import NUFFT_cpu
+
+except Exception:
+    warnings.warn("pyNUFFT python package has not been found. Exists only "
+                  "in Python3.x. Try Pynfft if non uniform Fourier"
+                  " transform needed")
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: Blank at the end.

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        return cls.numOfInstances
+    getNumInstances = classmethod(getNumInstances)
+
+    def __init__(self):
+        self.countInstances()
+
+
+class NUFFT(FourierBase, Singleton):
+    """  N-D non uniform Fast Fourrier Transform class
+
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: Put a note here on the mutliple intances problem and the link with the singleton inheritance.

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            string indicating which hardware platform will be used to
+            compute the NUFFT
+        Kd: int or tuple
+            int or tuple indicating the size of the frequency grid,
+            for regridding. If int, will be evaluated
+            to (Kd,)*nb_dim of the image
+        Jd: int or tuple
+            Size of the interpolator kernel. If int, will be evaluated
+            to (Jd,)*dims image
+
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: Empy line to remove.

File: pysap/plugins/mri/reconstruct_3D/fourier.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        cls.numOfInstances += 1
+
+    countInstances = classmethod(countInstances)
+
+    def getNumInstances(cls):
+        return cls.numOfInstances
+    getNumInstances = classmethod(getNumInstances)
+
+    def __init__(self):
+        self.countInstances()
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/fourier.py]: Why not using a metaclass? 

<pre>
from pisap.base.utils import with_metaclass

class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]

class NUFFT(with_metaclass(Singleton)):
    ...
</pre>

File: pysap/plugins/mri/reconstruct_3D/utils.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    dim1, dim2, dim3 = np.where(mask == 1)
+    dim1 = dim1.astype("float") / mask.shape[0] - 0.5
+    dim2 = dim2.astype("float") / mask.shape[1] - 0.5
+    dim3 = dim3.astype("float") / mask.shape[2] - 0.5
+    return np.c_[dim1, dim2, dim3]
+
+
+def normalize_samples(samples_locations):
+    """Normalize the 3D samples between [-.5; .5[
```
Comment (AGrigis) [inline pysap/plugins/mri/reconstruct_3D/utils.py]: Add white space.

File: pysap/plugins/mri/reconstruct_3D/fourier.py (hunk 5 above)
Comment (bensarthou) [reply pysap/plugins/mri/reconstruct_3D/fourier.py]: the one from Modopt has a nice color wrapping + doesnt copy the line of the warning. I can change if necessary

**Comment by AGrigis** [comment]
MRI plugins is now here: https://github.com/CEA-COSMIC/pysap-mri.git
Please resubmit your PR.

**Pull request closed without merging**